pkg/virt-handler/device-manager/device_controller.go

```diff
@@ -95,6 +97,8 @@ func (c *DeviceController) Run(stop chan struct{}) error {
 		go c.startDevicePlugin(dev, stop)
 	}
 
+	util.MarkReady(ReadinessFile)
```

Comment:
Could you explain how this is working? When looking at the snippes provided in the PR description it looks to me like our device plugin already reported the resources and that virt-handler marked the node as schedulable too. How does a readiness probe on virt-handler help mitigating this situation?

Reply:
What can happen is:  
   - virt-handler starts, it registers the device plugin 
   - virt-handler pods are deleted / restarted 
   -  virt-handler is ready but device plugin will not yet be re-registered 
This leaves the node to have the following state when the virt-hander is first running.. 
```  
allocatable:
    devices.kubevirt.io/kvm: "0"
    ... 
capacity:
    devices.kubevirt.io/kvm: "110"
    ... 
```

Since the pod currently has no readiness check we see it has available as soon as it is running but there is about 60 seconds from when it starts to the device manager running and register the plugins.

Reply:
What I don't understand: If there are no allocatable, why are pods scheduled there?

Reply:
"Update plugin resources failed due to requested..." is produced by Allocate of device plugin handler in kubelet. So it suggests that device plugin(virt-handler) was running(because we have resources on node) but is unable to serve resources at that moment.  
Do we do this  https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/pkg/apis/deviceplugin/v1alpha/api.proto#L43 ?

Reply:
What I mean:

```
allocatable:
    devices.kubevirt.io/kvm: "0"
    ... 
capacity:
    devices.kubevirt.io/kvm: "110"
    ... 
```

Looks correct. No pods should be scheduled to the node if it looks like this. Can it be that the kubelet only detects that right after one pod got there and requested it?

Reply:
I think so. Scheduler have own copy of the state so it can be(right?). Does the allocatable stays at 0? Why did it crash in first place?

Reply:
I've update this PR to instead update the `kubevirt.io/schedulable=false` label so that if the device plugin is not running the node is not schedulable. The event's in the case of a pod being created during a `virt-handler` pods restarting now look as follows

```
89s         Normal    SuccessfulCreate       virtualmachineinstance/testvmizpzf7f9fhhmcxwl9bj7bn5sl5clf9xg2csg9d4fztxc86jqh   Created virtual machine pod virt-launcher-testvmizpzf7f9fhhmcxwl9bj7bn5sl5clf9xg2csg9drz7ws
89s         Warning   FailedScheduling       pod/virt-launcher-testvmizpzf7f9fhhmcxwl9bj7bn5sl5clf9xg2csg9drz7ws              0/2 nodes are available: 2 node(s) didn't match node selector.
58s         Warning   FailedScheduling       pod/virt-launcher-testvmizpzf7f9fhhmcxwl9bj7bn5sl5clf9xg2csg9drz7ws              0/2 nodes are available: 1 Insufficient devices.kubevirt.io/kvm, 1 node(s) didn't match node selector.
58s         Warning   FailedScheduling       pod/virt-launcher-testvmizpzf7f9fhhmcxwl9bj7bn5sl5clf9xg2csg9drz7ws              0/2 nodes are available: 2 Insufficient devices.kubevirt.io/vhost-net.
53s         Warning   FailedScheduling       pod/virt-launcher-testvmizpzf7f9fhhmcxwl9bj7bn5sl5clf9xg2csg9drz7ws              0/2 nodes are available: 1 Insufficient devices.kubevirt.io/kvm, 1 Insufficient devices.kubevirt.io/tun.
56s         Warning   FailedScheduling       pod/virt-launcher-testvmizpzf7f9fhhmcxwl9bj7bn5sl5clf9xg2csg9drz7ws              0/2 nodes are available: 2 Insufficient devices.kubevirt.io/tun.
53s         Warning   FailedScheduling       pod/virt-launcher-testvmizpzf7f9fhhmcxwl9bj7bn5sl5clf9xg2csg9drz7ws              0/2 nodes are available: 1 Insufficient devices.kubevirt.io/kvm, 1 Insufficient devices.kubevirt.io/vhost-net.
```

The allocatable stays at 0 until the virt-handler re-registers the device manager, in this case the virt-handler was restarted because the test that runs before it updates the DS restarting the pods.

Reply:
Thanks, that looks better. Can you share which test? I feel like we can improve it so it doesn't disturb other tests.